### PR TITLE
Firewall executor

### PR DESCRIPTION
**Pull request by hehegrrrhehegrr**

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -8,3 +8,5 @@ Cargo.lock
 
 # These are backup files generated by rustfmt
 **/*.rs.bk
+
+/.vscode
```
Comment (stuffy-idler) [inline .gitignore:12]: Add newline at the end of all files, it's common best practise

File: firewall_redis/Cargo.toml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+[package]
+name = "firewall_redis"
+version = "0.1.0"
+edition = "2021"
+
+# See more keys and their definitions at https://doc.rust-lang.org/cargo/reference/manifest.html
+
+[dependencies]
+tokio = { version = "*", features = ["full"] }
```
Comment (stuffy-idler) [inline firewall_redis/Cargo.toml:9]: Tokio docs describes all of its features. Please, specify used  only

File: firewall_redis/config.yaml [hunk 3]
```diff
@@ -0,0 +1,5 @@
+redis:
+  conn_string: 'redis://127.0.0.1:6379'
```
Comment (stuffy-idler) [inline firewall_redis/config.yaml:2]: It's better to build conn string at the code, but not providing it via config.

File: firewall_redis/src/model.rs [hunk 4]
```diff
@@ -0,0 +1,92 @@
...
+#[derive(Debug, Serialize, Deserialize)]
+pub struct BanTarget {
+    #[serde(rename = "type")]
+    pub target_type: TargetType,
+
+    #[serde(rename = "value")]
+    pub value: String,
+}
+
+impl Debug for TargetType {
```
Comment (stuffy-idler) [inline firewall_redis/src/model.rs:33]: I think `Debug` deriving can give the same result and I suggest to use it.
If you want to update case of values while serialize/deserialize, you can specify `#[serde(rename_all = "snake_case")]` for entire struct

File: firewall_redis/src/model.rs [hunk 5]
```diff
@@ -0,0 +1,92 @@
...
+        })
+    }
+}
+
+#[derive(Debug, Serialize, Deserialize)]
+pub struct BanTarget {
+    #[serde(rename = "type")]
+    pub target_type: TargetType,
+
+    #[serde(rename = "value")]
```
Comment (stuffy-idler) [inline firewall_redis/src/model.rs:29]: It's excess, `value` is already `value`

File: firewall_redis/src/main.rs [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+
+    let json_cfg = web::JsonConfig::default()
+    .content_type(|mime| mime == mime::APPLICATION_JSON)
+    .error_handler(|err, req| {
+        error::InternalError::from_response(err, HttpResponse::BadRequest().into()).into()
+    });
+
+    let redis_pool_data = web::Data::new(pool);
+    HttpServer::new(move || {
+        App::new()
```
Comment (stuffy-idler) [inline firewall_redis/src/main.rs:50]: Let's move it to separate module with its child (process_ban)

File: firewall_redis/src/main.rs [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+        Some(s) => s.to_str().unwrap().to_string()
+    };
+    let ban = match BanEntity::from(ban_req.0, anl.clone()){
+        Ok(b) => b,
+        Err(fe) => return fe.into(),
+    };
+
+    let mut handles = vec![];
+
+    for bt in &ban.target{
```
Comment (stuffy-idler) [inline firewall_redis/src/main.rs:74]: Your purpose here is to transform `ban.target` to list of futures, so you can do it using `map`

File: firewall_redis/src/main.rs [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+        let pool = pool.clone();
+        let key = format!("{}", bt.value);
+
+        let anl = anl.clone();
+        let reason = ban.reason.clone();
+        let ttl = ban.ttl.clone();
+
+        handles.push(tokio::spawn(async move {
```
Comment (stuffy-idler) [inline firewall_redis/src/main.rs:82]: It doesn't needed to do value fetching at separate thread, let's make it simpler with a basic future

File: firewall_redis/src/main.rs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            let mut conn = match pool.get().await {
+                Ok(c) => c,
+                Err(e) => {error!(""); return Err(e.to_string())}
```
Comment (stuffy-idler) [inline firewall_redis/src/main.rs:85]: You have an error `e`, so you can print it. Empty string is not descriptive what is problem, try not to do things like this one

File: firewall_redis/src/main.rs [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+    for bt in &ban.target{
+        let pool = pool.clone();
+        let key = format!("{}", bt.value);
+
+        let anl = anl.clone();
+        let reason = ban.reason.clone();
+        let ttl = ban.ttl.clone();
+
+        handles.push(tokio::spawn(async move {
+            let mut conn = match pool.get().await {
```
Comment (stuffy-idler) [inline firewall_redis/src/main.rs:83]: If you are not going to do anything with value from future, but to do smth with an error, it's better to:
```rust
let mut conn = match pool.get().await.map_err(|e| {
  error!("{:?}", e);
  Err(e.to_string())
})?;

File: firewall_redis/src/main.rs [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+                Ok(c) => c,
+                Err(e) => {error!(""); return Err(e.to_string())}
+            };
+    
+            match cmd("HSET").arg(&key).arg(anl).arg(reason).query_async(&mut *conn).await
```
Comment (stuffy-idler) [inline firewall_redis/src/main.rs:88]: In this case you are not doing anything with future result, only err, so you can do:
```rust
cmd("HSET")
  .arg(&key)
  .arg(anl)
  .arg(reason)
  .query_async(&mut *conn)
  .await
  .map_err(|re| re.to_string())?;
```

File: firewall_redis/src/main.rs [hunk 12]
```diff
@@ -0,0 +1,117 @@
...
+                as Result<(), RedisError> {
+                Err(re) => return Err(re.to_string()),
+                _ => (),
+            };
+
+            match cmd("EXPIRE").arg(&key).arg(ttl).arg("GT")
```
Comment (stuffy-idler) [inline firewall_redis/src/main.rs:94]: The same notice as above

File: firewall_redis/src/main.rs [hunk 13]
```diff
@@ -0,0 +1,117 @@
...
+                .query_async(&mut *conn).await
+                as Result<(), RedisError> {
+                Err(re) => return Err(re.to_string()),
+                _ => (),
+            };
+            Ok(())
+        }));
+    }
+
+    for res in join_all(handles).await {
```
Comment (stuffy-idler) [inline firewall_redis/src/main.rs:104]: Rewrite it using combinators, it will be more readable and supportable

File: firewall_redis/src/main.rs [hunk 14]
```diff
@@ -0,0 +1,117 @@
...
+async fn main() -> std::io::Result<()> {
+    info!("start application");
+
+    let cfg: config::Config = match pepe_config::load(DEFAULT_CONFIG, ::config::FileFormat::Yaml) {
+        Ok(a) => a,
+        Err(e) => panic!("can't read config {:?}", e)
+    };
+    info!("config loaded"; "config" => &cfg);
+
+    let rcm = match RedisConnectionManager::new(cfg.redis.conn_string.to_string()) {
```
Comment (stuffy-idler) [inline firewall_redis/src/main.rs:33]: Let's move redis pool initialization to separate module

File: firewall_redis/src/config.rs [hunk 15]
```diff
@@ -0,0 +1,13 @@
+use serde::{Deserialize, Serialize};
+use slog_extlog_derive::SlogValue;
+
+use crate::redis;
+use crate::server;
+
+#[derive(Clone, Debug, Serialize, Deserialize, SlogValue)]
+pub struct Config {
+    #[serde(rename(deserialize = "redis"))]
```
Comment (stuffy-idler) [inline firewall_redis/src/config.rs:9]: It does not need to be renamed (`server` too)

**Review by stuffy-idler — CHANGES_REQUESTED** [review]

File: firewall_redis/src/main.rs (hunk 13 above)
Comment (hehegrrrhehegrr) [reply firewall_redis/src/main.rs]: Tried to put everything into one combinator, looked cumbersome and unreadable, so let's keep it as it is now - 2 separate combinators.

**Review by stuffy-idler — CHANGES_REQUESTED** [review]

File: firewall_redis/src/redis.rs [hunk 16]
```diff
@@ -16,14 +25,76 @@ impl Config {
     }
 }
 
-pub async fn get_pool(cfg: &Config) -> Result<Pool<RedisConnectionManager>, String> {
-    let rcm = match RedisConnectionManager::new(cfg.connection_string()) {
-        Ok(c) => c,
-        Err(re) => return Err(format!("can't setup redis: {:?}", re.to_string())),
-    };
+impl Service {
+    pub async fn new<'a>(cfg: &Config) -> Result<Self, String> {
```
Comment (stuffy-idler) [inline firewall_redis/src/redis.rs:29]: Return `String` is not informative, better to return concrete Error.
There is best practice to do this: create `enum Error` with derived `thiserror::Error` with all of failing cases, and use it as with `Result`:
```rust
use error::Error as AppError;

// ...

pub async fn new<'a>(cfg: &Config) -> Result<Self, AppError> {
  // ...
}

// match bb8::Pool::builder().build(rcm).await {
//   Err(e) => Err(format!("build pool: {:?}", e.to_string())),
//   Ok(p) => Ok(p),
// }

bb8::Pool::builder()
  .build(rcm)
  .await
  .map_err(|err| AppError::Bb8PoolBuild(err))
```

File: firewall_redis/src/server.rs [hunk 17]
```diff
@@ -1,8 +1,85 @@
...
 pub struct Config {
     pub host: String,
     pub port: u16,
 }
+
+pub struct Server {
+    ws_cfg: web::ServiceConfig,
+}
+
+fn configure(cfg: &mut web::ServiceConfig, red: redis::Service) {
```
Comment (stuffy-idler) [inline firewall_redis/src/server.rs:19]: Mutating is not true rust way to write the code. It's better to create new instance of smth from provided data

**Review by stuffy-idler — CHANGES_REQUESTED** [review]

**Pull request closed without merging**